File: backend/src/main/java/cn/etarch/mao/harness/core/SessionHistoryLoader.java
package cn.etarch.mao.harness.core;

import cn.etarch.mao.harness.llm.ChatRequest;
import cn.etarch.mao.session.entity.Message;
import cn.etarch.mao.session.service.SessionService;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.List;

/**
 * 会话历史加载与应用到执行上下文（纯加载/应用，无压缩副作用）。
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class SessionHistoryLoader {

    private final SessionService sessionService;
    private final ContextManager contextManager;
    private final ObjectMapper objectMapper;

    public HistorySnapshot loadHistoryAfterBoundary(Long sessionId, long boundary) {
        List<Message> rawMessages = sessionService.getMessagesAfterId(sessionId, boundary);
        List<Long> snapshotMessageIds = rawMessages.stream().map(Message::getId).toList();
        List<Message> normalized = MessageHistoryNormalizer.normalizeEntities(rawMessages, objectMapper);
        List<PersistedChatMessage> persistedMessages = normalized.stream()
                .map(message -> new PersistedChatMessage(
                        message.getId(), message.getContent(), toChatMessage(message)))
                .toList();
        return new HistorySnapshot(snapshotMessageIds, normalized, persistedMessages);
    }

    /**
     * 用摘要 + DB 增量替换 context.messages（clear+addAll，不替换列表引用），
     * 并按原顺序复原内存态 system 消息（后台任务结果 / MCP 降级提示等）。
     */
    public void applyHistory(AgentExecutionContext context, String summary, HistorySnapshot history) {
        List<ChatRequest.Message> incrementalMessages = history.persistedMessages().stream()
                .map(PersistedChatMessage::chatMessage)
                .toList();
        context.getMessages().clear();
        context.getMessages().addAll(contextManager.prependSessionSummary(summary, incrementalMessages));
        // 复原从未落库的 ephemeral system 消息（语义上位于列表尾部）
        if (context.getEphemeralSystemMessages() != null && !context.getEphemeralSystemMessages().isEmpty()) {
            context.getMessages().addAll(context.getEphemeralSystemMessages());
        }
        context.setSessionSummary(summary);
        context.getToolAttachments().clear();
        context.getToolAttachments().putAll(
                ToolAttachmentLoader.loadAllFromMessages(history.normalizedEntities(), objectMapper));
    }

    public ChatRequest.Message toChatMessage(Message message) {
        var builder = ChatRequest.Message.builder()
                .role(message.getRole().toLowerCase())
                .content(parseContent(message.getContent()));
        if (message.getToolCallId() != null) {
            builder.toolCallId(message.getToolCallId());
        }
        if (message.getToolCalls() != null && !message.getToolCalls().isEmpty()) {
            try {
                builder.toolCalls(objectMapper.readValue(
                        message.getToolCalls(), new TypeReference<List<ChatRequest.ToolCall>>() {}));
            } catch (JsonProcessingException e) {
                log.warn("Failed to parse tool_calls for message {}", message.getId(), e);
            }
        }
        return builder.build();
    }

    private Object parseContent(String raw) {
        if (raw == null) return "";
        String trimmed = raw.trim();
        if (trimmed.startsWith("[")) {
            try {
                return objectMapper.readValue(trimmed,
                        new TypeReference<List<ChatRequest.ContentPart>>() {});
            } catch (Exception e) {
                return raw;
            }
        }
        return raw;
    }

    public record HistorySnapshot(
            List<Long> snapshotMessageIds,
            List<Message> normalizedEntities,
            List<PersistedChatMessage> persistedMessages) {
    }
}

File: backend-ts/src/harness/core/session-history-loader.ts
import type { ChatMessage, ChatRequest, ToolCall } from '../llm/chat-request.js';
import type { Message, SessionService } from '../deps.js';
import type { AgentExecutionContext } from './agent-execution-context.js';
import type { ContextManager } from './context-manager.js';
import { MessageHistoryNormalizer } from './message-history-normalizer.js';
import { PersistedChatMessage } from './persisted-chat-message.js';
import { ToolAttachmentLoader } from './tool-attachment-loader.js';
import { harnessLog } from '../log.js';

export interface HistorySnapshot {
  snapshotMessageIds: number[];
  normalizedEntities: Message[];
  persistedMessages: PersistedChatMessage[];
}

export class SessionHistoryLoader {
  constructor(
    private readonly sessionService: SessionService,
    private readonly contextManager: ContextManager,
  ) {}

  async loadHistoryAfterBoundary(sessionId: number, boundary: number): Promise<HistorySnapshot> {
    const rawMessages = await this.sessionService.getMessagesAfterId(sessionId, boundary);
    const snapshotMessageIds = rawMessages.map((m) => m.id!);
    const normalized = MessageHistoryNormalizer.normalizeEntities(rawMessages, parseToolCallsJson) ?? rawMessages;
    const persistedMessages = normalized.map((message) => new PersistedChatMessage(
      message.id!,
      message.content ?? '',
      this.toChatMessage(message),
    ));
    return { snapshotMessageIds, normalizedEntities: normalized, persistedMessages };
  }

  applyHistory(context: AgentExecutionContext, summary: string | null | undefined, history: HistorySnapshot): void {
    const incremental = history.persistedMessages.map((p) => p.chatMessage);
    context.messages.length = 0;
    context.messages.push(...this.contextManager.prependSessionSummary(summary, incremental));
    if (context.ephemeralSystemMessages.length > 0) {
      context.messages.push(...context.ephemeralSystemMessages);
    }
    context.sessionSummary = summary;
    context.toolAttachments.clear();
    for (const [k, v] of Object.entries(ToolAttachmentLoader.loadAllFromMessages(history.normalizedEntities))) {
      context.toolAttachments.set(k, v);
    }
  }

  toChatMessage(message: Message): ChatMessage {
    const msg: ChatMessage = {
      role: message.role?.toLowerCase(),
      content: parseContent(message.content),
    };
    if (message.toolCallId) msg.toolCallId = message.toolCallId;
    if (message.toolCalls) {
      try {
        msg.toolCalls = parseToolCallsJson(message.toolCalls);
      } catch (e) {
        harnessLog('warn', `Failed to parse tool_calls for message ${message.id}`, e);
      }
    }
    // DeepSeek thinking 模式要求带 tools 的多轮请求回传历史 assistant 的 reasoning_content，
    // 崩溃恢复 / compaction 重载历史时需从持久化的 thinking_content 还原，否则后续请求 400。
    if (message.thinkingContent) msg.reasoningContent = message.thinkingContent;
    return msg;
  }
}

function parseToolCallsJson(json: string): ToolCall[] {
  try {
    const parsed = JSON.parse(json) as unknown;
    return Array.isArray(parsed) ? parsed as ToolCall[] : [];
  } catch {
    return [];
  }
}

function parseContent(raw: string | null | undefined): unknown {
  if (raw == null) return '';
  const trimmed = raw.trim();
  if (trimmed.startsWith('[')) {
    try {
      const parsed = JSON.parse(trimmed) as unknown;
      // 只有每个元素都是含 type 的多模态 content part 才按数组解析；
      // 用户发送的纯文本 JSON 数组（如 [1,2,3]、["a","b"]、[]）在 DB 中与多模态数组无法区分，
      // 若不校验结构会被误当作 content part 数组发给 LLM，产生 400 或内容损毁。
      if (isMultimodalContentParts(parsed)) return parsed;
      return raw;
    } catch {
      return raw;
    }
  }
  return raw;
}

/** 每个元素必须是含 type（text/image_url）的对象，且任一元素不满足则整体按原始字符串返回。 */
function isMultimodalContentParts(parsed: unknown): parsed is Array<Record<string, unknown>> {
  if (!Array.isArray(parsed) || parsed.length === 0) return false;
  return parsed.every((part) => {
    if (part == null || typeof part !== 'object' || Array.isArray(part)) return false;
    const map = part as Record<string, unknown>;
    return map.type === 'text' || map.type === 'image_url';
  });
}
